docs/testing/e2e-plan.md

```diff
@@ -0,0 +1,79 @@
+
+[^1]: Early E2E tests were run against production artifacts, generally because they were the only reliably accessible artifacts at the time.
+
+## The End-Goal
+
+To be able to provide confidence that the CP offerings are working across our supported EKS-A versions, EKS-A providers, cluster node OSes, and K8s versions, the current framework must be improved in the following ways:
+
++ **Artifact version control:**
+  Tests should be run against an arbitrary versions of artifacts, defaulting to their latest versions.
+
++ **Timing:**
+  E2E tests should be run on every revision of an "ok to test" PR. This may present some challenges for the current E2E test architecture, as it might require communication between separate AWS pipelines, or failing that, custom code that can communicate with other AWS resources to identify and list available artifacts built in other pipelines.
+
++ **Platform configuration:**
+  A test should be able to exclude specific EKS-A versions, EKS-A providers, cluster node OSes, or K8s versions with which they are incompatible or unsupported. A test will be run on every permutation by default.
+
++ **Configuration review & control:**
+  The test dimensions should be easily viewed, verified, and edited by humans (or reasonable facsimiles thereof). This provides a means of quickly excluding tests categorically for reasons such as hardware outage.
+
+Additionally, it is foreseen that the number of E2E tests that need to be run for each PR will quickly grow, and so it is recommended that these performance goals also be considered:
+
++ **Reusable clusters:** The amount of time required to create and destroy clusters before and after each test will quickly become overwhelming with reasonable amounts of hardware.
+
+## The Path Forward
+
+### Short Term
+
+#### Test Against Current Branch
+
+The current tests will be modified to run against the latest versions of artifacts. This will be accomplished by removing the existing hardcoded registry links from the existing tests, and implementing a means by which URLs to test-run-specific artifacts will be used. These up-to-date artifacts will be used (at a minimum) when installing the EKS-A Curated Packages controller during cluster creation, as well as when said controller acquires its active bundle. The newly-acquired active bundle will likewise contain links to newly-built test artifacts, such that the end-to-end tests will be run against artifacts built using the current git commits under test, and not the hard-coded development or production registries, as is the case now.
```

Comment:
Will there be a way for a developer to specify their build of package controller to run the local E2E test as an option? Or is the expectation that this will be done in the long term with running e2e test on ok-to-test pull requests?

Reply:
In the short term, the plan is to simply run the tests against dev versions via the existing build server. The long term plan includes allowing local runs against arbitrary versions of packages via a custom manifest file.

Depending on how the variables for the build server are passed, it is possible that the ability to specify a local build for local testing would be possible, but I haven't confirmed that, and I'd consider it a bonus, as opposed to a stated goal.